boot/bootutil/src/loader.c

```diff
@@ -793,58 +855,122 @@ boot_copy_sector(int flash_area_id_src, int flash_area_id_dst,
 static void
 boot_swap_sectors(int idx, uint32_t sz, struct boot_status *bs)
 {
+    const struct flash_area *fap;
     uint32_t copy_sz;
+    uint32_t trailer_sz;
     uint32_t img_off;
+    uint32_t scratch_trailer_off;
+    struct boot_swap_state swap_state;
     int rc;
 
     /* Calculate offset from start of image area. */
     img_off = boot_data.imgs[0].sectors[idx].fa_off -
               boot_data.imgs[0].sectors[0].fa_off;
 
+    copy_sz = sz;
+
+    trailer_sz = boot_slots_trailer_sz(boot_data.write_sz);
+    if (boot_data.imgs[0].sectors[idx].fa_off + sz >
+        boot_data.imgs[0].sectors[boot_data.imgs[0].num_sectors - 1].fa_off) {
+        copy_sz -= trailer_sz;
+    }
+
+    bs->use_scratch = (bs->idx == 0 && copy_sz != sz);
+
     if (bs->state == 0) {
         rc = boot_erase_sector(FLASH_AREA_IMAGE_SCRATCH, 0, sz);
         assert(rc == 0);
 
         rc = boot_copy_sector(FLASH_AREA_IMAGE_1, FLASH_AREA_IMAGE_SCRATCH,
-                              img_off, 0, sz);
+                              img_off, 0, copy_sz);
         assert(rc == 0);
 
+        if (bs->idx == 0) {
```

Comment:
Your analysis looks good but seems to be missing the `if (bs->use_scratch)` tests. `boot_status_init_by_id` initializes a new trailer from an existing one, exactly as you said. But it chooses which one to initialize based on the fact that we are copying the last sector. If we are (as is always the case for stm32f401) it should not erase SLOT0 and instead use scratch for writing a temporary trailer that will only end up in SLOT0 when state is 2.

Btw, I have not really tested on the stm32f401 (only stm32f407) but would be great to test it to be sure if this really works. Unfortunately I don't have the board here with me but I'll be at Runtime's HQ by Thursday and can give it a try.